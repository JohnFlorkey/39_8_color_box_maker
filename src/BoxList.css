.BoxList {
    display: flex;
    flex-wrap: row;
}