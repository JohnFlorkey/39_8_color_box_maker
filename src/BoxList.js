import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';
import './BoxList.css';

function BoxList() {
	const initialState = [
		{backgroundColor: "red", width: 200, height: 200, id: uuid()},
		{backgroundColor: "blue", width: 100, height: 100, id: uuid()}
	]

	const [ boxes, setBoxes ] = useState(initialState);

	function addBox(backgroundColor, width, height) {
		setBoxes(boxes => [...boxes, {backgroundColor, width: `${width}px`, height: `${height}px`, id: uuid()}]);
	}

	function removeBox(id) {
		setBoxes(boxes => boxes.filter(b => b.id !== id));
	}

	return (
		<div>
			<div className="BoxList">
				{boxes.map(b => <Box key={b.id} backgroundColor={b.backgroundColor} width={b.width} height={b.height} removeBox={() => removeBox(b.id)}/>)}
			</div>
			<NewBoxForm addBox={addBox} />
		</div>
		
	);
}

export default BoxList;