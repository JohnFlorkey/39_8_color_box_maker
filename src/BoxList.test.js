import { fireEvent, render } from '@testing-library/react';
import BoxList from './BoxList';

// smoke test
it("should render without error", () => {
    render(<BoxList />);
});

// snapshot test
it("should match the snapshot", () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

// business logic tests
/** kind of weird, the react-testing-library philosophy is to select elements from
 *  the DOM based on text or a value that the user can see. This example app doesn't
 * render much text, so the options for elements to select are limited.
 */

it("should add box", () => {
    const {getByText, getByLabelText, queryByText} = render(<BoxList />);
    
    expect(queryByText("X")).not.toBeInTheDocument();

    const backgroundColor = getByLabelText("Background Color");
    const width = getByLabelText("Width");
    const height = getByLabelText("Height");
    const addBtn = getByText("Add Box");

    fireEvent.change(backgroundColor, { target: { value: "red"}});
    fireEvent.change(width, { target: { value: "200"}});
    fireEvent.change(height, { target: { value: "150"}});
    fireEvent.click(addBtn);

    const removeBtn = getByText("X")
    expect(removeBtn).toBeInTheDocument(); // button rendered with the colored div
})

it("should remove the box", () => {
    const {getByText, getByLabelText, queryByText} = render(<BoxList />);
    
    expect(queryByText("X")).not.toBeInTheDocument();

    const backgroundColor = getByLabelText("Background Color");
    const width = getByLabelText("Width");
    const height = getByLabelText("Height");
    const addBtn = getByText("Add Box");

    fireEvent.change(backgroundColor, { target: { value: "red"}});
    fireEvent.change(width, { target: { value: "200"}});
    fireEvent.change(height, { target: { value: "150"}});
    fireEvent.click(addBtn);

    const removeBtn = getByText("X")
    expect(removeBtn).toBeInTheDocument(); // button rendered with the colored div

    fireEvent.click(removeBtn);
    expect(queryByText("X")).not.toBeInTheDocument();
})