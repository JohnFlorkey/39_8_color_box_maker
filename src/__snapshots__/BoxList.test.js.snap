// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="BoxList"
    />
    <form
      class="NewBoxForm"
    >
      <label
        class="NewBoxForm-label"
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        class="NewBoxForm-input"
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <label
        class="NewBoxForm-label"
        for="width"
      >
        Width
      </label>
      <input
        class="NewBoxForm-input"
        id="width"
        name="width"
        type="text"
        value=""
      />
      <label
        class="NewBoxForm-label"
        for="height"
      >
        Height
      </label>
      <input
        class="NewBoxForm-input"
        id="height"
        name="height"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
